src/pybind/mgr/cephadm/upgrade.py

```diff
@@ -119,29 +119,23 @@ def continue_upgrade(self) -> bool:
             return True
         return False
 
-    def _wait_for_ok_to_stop(self, s) -> bool:
+    def _wait_for_ok_to_stop(self, s: DaemonDescription) -> bool:
         # only wait a little bit; the service might go away for something
         tries = 4
         while tries > 0:
-            if s.daemon_type not in ['mon', 'osd', 'mds']:
-                logger.info('Upgrade: It is presumed safe to stop %s.%s' %
-                              (s.daemon_type, s.daemon_id))
-                return True
-            ret, out, err = self.mgr.mon_command({
-                'prefix': '%s ok-to-stop' % s.daemon_type,
-                'ids': [s.daemon_id],
-            })
             if not self.upgrade_state or self.upgrade_state.get('paused'):
                 return False
-            if ret:
-                logger.info('Upgrade: It is NOT safe to stop %s.%s' %
-                              (s.daemon_type, s.daemon_id))
-                time.sleep(15)
-                tries -= 1
-            else:
-                logger.info('Upgrade: It is safe to stop %s.%s' %
+
+            ok = self.mgr.cephadm_services[s.daemon_type].ok_to_stop([s.daemon_id])
```

Comment:
What if daemon_type is not contained in cephadm_services?

Reply:
then this is going to break hard. maybe we can use a mypy extension to statically verify this. Like https://mypy.readthedocs.io/en/stable/more_types.html#typeddict or something. This would tie us to mypy.

Reply:
The static verification would be nice, but I'm not certain we want to add tie this to mypy.